abci/check_tx.go

```diff
@@ -0,0 +1,214 @@
+			}
+		}()
+
+		tx, err := handler.txDecoder(req.Tx)
+		if err != nil {
+			return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("failed to decode tx: %w", err), 0, 0, nil, false)
+		}
+
+		// Attempt to get the bid info of the transaction.
+		bidInfo, err := handler.mempool.GetAuctionBidInfo(tx)
+		if err != nil {
+			return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("failed to get auction bid info: %w", err), 0, 0, nil, false)
+		}
+
+		// If this is not a bid transaction, we just execute it normally.
+		if bidInfo == nil {
+			return handler.baseApp.CheckTx(req)
+		}
+
+		// We attempt to get the latest committed state in order to verify transactions
+		// as if they were to be executed at the top of the block. After verification, this
+		// context will be discarded and will not apply any state changes.
+		ctx := handler.GetContextForBidTx(req)
+
+		// Verify the bid transaction.
+		gasInfo, err := handler.ValidateBidTx(ctx, tx, bidInfo)
+		if err != nil {
+			return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("invalid bid tx: %w", err), gasInfo.GasWanted, gasInfo.GasUsed, nil, false)
+		}
+
+		// If the bid transaction is valid, we know we can insert it into the mempool for consideration in the next block.
+		if err := handler.mempool.Insert(ctx, tx); err != nil {
+			return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("invalid bid tx; failed to insert bid transaction into mempool: %w", err), gasInfo.GasWanted, gasInfo.GasUsed, nil, false)
+		}
+
+		return cometabci.ResponseCheckTx{
+			Code:      cometabci.CodeTypeOK,
+			GasWanted: int64(gasInfo.GasWanted),
+			GasUsed:   int64(gasInfo.GasUsed),
+		}
+	}
+}
+
+// ValidateBidTx is utilized to verify the bid transaction against the latest committed state.
+func (handler *CheckTxHandler) ValidateBidTx(ctx sdk.Context, bidTx sdk.Tx, bidInfo *mempool.AuctionBidInfo) (sdk.GasInfo, error) {
+	// Verify the bid transaction.
+	ctx, err := handler.anteHandler(ctx, bidTx, false)
+	if err != nil {
+		return sdk.GasInfo{}, fmt.Errorf("invalid bid tx; failed to execute ante handler: %w", err)
+	}
+
+	// Store the gas info and priority of the bid transaction before applying changes with other transactions.
+	gasInfo := sdk.GasInfo{
+		GasWanted: ctx.GasMeter().Limit(),
+		GasUsed:   ctx.GasMeter().GasConsumed(),
+	}
+
+	// Verify all of the bundled transactions.
+	for _, tx := range bidInfo.Transactions {
+		bundledTx, err := handler.mempool.WrapBundleTransaction(tx)
+		if err != nil {
+			return gasInfo, fmt.Errorf("invalid bid tx; failed to decode bundled tx: %w", err)
+		}
+
+		bidInfo, err := handler.mempool.GetAuctionBidInfo(bundledTx)
+		if err != nil {
+			return gasInfo, fmt.Errorf("invalid bid tx; failed to get auction bid info: %w", err)
+		}
+
+		// Bid txs cannot be included in bundled txs.
+		if bidInfo != nil {
+			return gasInfo, fmt.Errorf("invalid bid tx; bundled tx cannot be a bid tx")
+		}
```

Comment:
Shouldn't the builder decorators check this?

Reply:
I think we did check this in the decorator but at some point we removed it (for performance reasons discussed with bera folks). IMO I think it does make sense to have this particular check in the builder decorator.

Reply:
Do you think should we add it back?